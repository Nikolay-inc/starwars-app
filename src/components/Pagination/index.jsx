'use client';

import { pageHandler } from '../../helpers/pageHandler';
import { ChakraProvider, Button, Flex } from '@chakra-ui/react';
import { Link } from '@chakra-ui/next-js';

// to paginate heroes list
const Pagination = ({ previousPage, nextPage }) => {
  // get number for next and prev pages
  const prevPageNumber = pageHandler(previousPage);
  const nextPageNumber = pageHandler(nextPage);

  return (
    <ChakraProvider>
      <Flex mt='30px' align='center' justify='flex-end' gap='30px'>
        {/* using chakra Link instead of the Next.js ones */}
        <Link href={`/?page=${prevPageNumber}`}>
          <Button 
            size='sm' 
            fontWeight='lite' 
            colorScheme='gray'
            isDisabled={!previousPage}
          >
            Previous Page
          </Button>
        </Link>
        <Link href={`/?page=${nextPageNumber}`}>
          <Button 
            size='sm' 
            fontWeight='lite' 
            colorScheme='gray'
            isDisabled={!nextPage}
          >
            Next Page
          </Button>
        </Link>
      </Flex>
    </ChakraProvider>
  );
}
  
export default Pagination;