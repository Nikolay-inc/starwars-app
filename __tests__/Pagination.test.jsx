import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Pagination from '../src/components/Pagination';

// mocking pageHandler function
jest.mock('../src/helpers/pageHandler', () => ({
  pageHandler: jest.fn((toPage) => {
    if (!toPage) return '';

    const newCurrentPage = new URLSearchParams(toPage.split('?')[1]).get('page');
    return newCurrentPage ? +newCurrentPage : 1;
  })
}));

describe('Pagination component', () => {
  test('renders Previous Page button', () => {
    render(<Pagination previousPage="https://example.com/?page=1" nextPage={"https://example.com/?page=2"} />);
    const prevPageButton = screen.getByRole('button', { name: /Previous Page/i });
    expect(prevPageButton).toBeInTheDocument();
  });

  test('renders Next Page button', () => {
    render(<Pagination previousPage="https://example.com/?page=1" nextPage={"https://example.com/?page=2"} />);
    const nextPageButton = screen.getByRole('button', { name: /Next Page/i });
    expect(nextPageButton).toBeInTheDocument();
  });

  test('disables Previous Page button when there is no previous page', () => {
    render(<Pagination />);
    const prevPageButton = screen.getByRole('button', { name: /Previous Page/i });
    expect(prevPageButton).toBeDisabled();
  });

  test('disables Next Page button when there is no next page', () => {
    render(<Pagination />);
    const nextPageButton = screen.getByRole('button', { name: /Next Page/i });
    expect(nextPageButton).toBeDisabled();
  });

  test('handles click on Previous Page button', () => {
    const mockPrevPageHandler = jest.fn();
    render(<Pagination previousPage="https://example.com/?page=1" nextPage={"https://example.com/?page=2"} />);
    const prevPageButton = screen.getByRole('button', { name: /Previous Page/i });
    userEvent.click(prevPageButton);
    expect(mockPrevPageHandler).toHaveBeenCalledTimes(0);
  });

  test('handles click on Next Page button', () => {
    const mockNextPageHandler = jest.fn();
    render(<Pagination previousPage={"https://example.com/?page=1"} nextPage="https://example.com/?page=3" />);
    const nextPageButton = screen.getByRole('button', { name: /Next Page/i });
    userEvent.click(nextPageButton);
    expect(mockNextPageHandler).toHaveBeenCalledTimes(0);
  });

  test('matches snapshot', () => {
    const { container } = render(
      <Pagination previousPage={"https://example.com/?page=1"} nextPage="https://example.com/?page=3" />
    );
    expect(container).toMatchSnapshot();
  });
});
