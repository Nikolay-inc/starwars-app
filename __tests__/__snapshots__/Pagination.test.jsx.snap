// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component matches snapshot 1`] = `
<div>
  <div
    class="css-vdprmx"
  >
    <a
      class="chakra-link css-spn4bz"
      href="/?page=1"
    >
      <button
        class="chakra-button css-24iywp"
        type="button"
      >
        Previous Page
      </button>
    </a>
    <a
      class="chakra-link css-spn4bz"
      href="/?page=3"
    >
      <button
        class="chakra-button css-24iywp"
        type="button"
      >
        Next Page
      </button>
    </a>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
